src/wp-includes/class-wp-block-scanner.php

```diff
@@ -0,0 +1,1244 @@
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $next_token_at = 0;
+
+	/**
+	 * Byte offset where namespace span begins.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $namespace_at = 0;
+
+	/**
+	 * Byte length of namespace span, or `0` if implicitly in the "core" namespace.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $namespace_length = 0;
+
+	/**
+	 * Byte offset where block name span begins.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $name_at = 0;
+
+	/**
+	 * Byte length of block name span.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $name_length = 0;
+
+	/**
+	 * Whether the delimiter contains the block-closing flag.
+	 *
+	 * This may be erroneous if present within a void block,
+	 * therefore the {@see self::has_closing_flag} can be used by
+	 * calling code to perform appropriate error-handling.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var bool
+	 */
+	private $has_closing_flag = false;
+
+	/**
+	 * Byte offset where JSON attributes span begins.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $json_at;
+
+	/**
+	 * Byte length of JSON attributes span, or `0` if none are present.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $json_length = 0;
+
+	/**
+	 * Internal parser state, differentiating whether the instance is currently matched,
+	 * on an implicit freeform node, in error, or ready to begin parsing.
+	 *
+	 * @see self::READY
+	 * @see self::MATCHED
+	 * @see self::HTML_OPEN
+	 * @see self::HTML_CLOSE
+	 * @see self::INCOMPLETE_INPUT
+	 * @see self::COMPLETE
+	 *
+	 * @since 6.9.0}
+	 *
+	 * @var string
+	 */
+	protected $state = self::READY;
+
+	/**
+	 * Indicates what kind of block comment delimiter this represents.
+	 *
+	 * One of:
+	 *
+	 *  - {@see static::OPENER} If the delimiter is opening a block.
+	 *  - {@see static::CLOSER} If the delimiter is closing an open block.
+	 *  - {@see static::VOID}   If the delimiter represents a void block with no inner content.
+	 *
+	 * If a parsed comment delimiter contains both the closing and the void
+	 * flags then it will be interpreted as a void block to match the behavior
+	 * of the official block parser, however, this is a mistake and probably
+	 * the block ought to close an open block of the same name, if one is open.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var string
+	 */
+	private $type;
+
+	/**
+	 * Creates a new block scanner.
+	 *
+	 * Example:
+	 *
+	 *     $scanner = WP_Block_Scanner::create( $html );
+	 *     while ( $scanner->next_delimiter() ) {
+	 *          if ( $scanner->opens_block( 'core/image' ) ) {
+	 *             echo "Found an image!\n";
+	 *         }
+	 *     }
+	 *
+	 * @see self::next_delimiter
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param string $source_text Input document potentially containing block content.
+	 * @return ?self Created block scanner, if successfully created.
+	 */
+	public static function create( string $source_text ) {
+		return new self( $source_text );
```

Comment:
Here, I think we want `static` to return a subclass:

```suggestion
		return new static( $source_text );
```

[Demo:](https://3v4l.org/3tTh6)

```php
<?php declare( strict_types = 1 );

class C {
    public static function create_self(): ?self {
        return new self();
    }
    public static function create_static(): ?static {
        return new static();
    }
}

class D extends C {}

var_dump(
    D::create_self(),
    D::create_static(),
);
/* prints:
object(C)#1 (0) {
}
object(D)#2 (0) {
}
*/
```

Reply:
this method is now gone.